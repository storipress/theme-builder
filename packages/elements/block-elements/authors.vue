<script lang="ts">
import type { PropType } from 'vue'
import Vue from 'vue'

import LinkElement from '../common/link-element.vue'
import { styleProps } from '../utils/style-props'
import TextElement from './text-element.vue'

interface Author {
  name: string
  url: string
}

export default Vue.extend({
  components: { LinkElement, TextElement },

  props: {
    kind: {
      type: String,
      required: true,
    },
    blockType: {
      type: String,
    },
    component: {
      type: String,
      default: 'p',
    },
    authorComponent: {
      type: String,
      default: 'span',
    },
    authors: {
      type: Array as PropType<Author[]>,
      required: true,
    },
    separator: {
      type: String,
      default: ', ',
    },
    authorClass: {
      type: String,
    },
    separatorClass: {
      type: String,
    },
    ...styleProps,
  },

  computed: {
    textProps() {
      const { component, authorComponent, authors, authorClass, sperataor, separatorClass, ...rest } = this.$props
      return {
        component: authorComponent,
        ...rest,
      }
    },
  },
})
</script>

<template>
  <component :is="component" class="flex items-center gap-x-1">
    <template v-for="(author, i) of authors">
      <template v-if="i === 0" /><span v-else-if="i === authors.length - 1" :key="`${i}-sep`" :class="separatorClass">
        and </span
      ><span v-else :key="`${i}-sep`" :class="separatorClass">{{ separator }}</span>
      <LinkElement :key="i" :href="author.url" :class="authorClass">
        <TextElement v-bind="textProps">{{ author.name }}</TextElement>
      </LinkElement>
    </template>
  </component>
</template>
