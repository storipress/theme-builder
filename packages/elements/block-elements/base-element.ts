import type { VueConstructor } from 'vue'
import type { Ref } from 'vue-demi'
import type { CommonInjected } from '../common/common-element'
import type { Article } from './types'
import { constant, noop } from 'lodash'

import Vue from 'vue'
import { inject, ref } from 'vue-demi'

type Maybe<T> = T | null

export interface ElementInfo {
  path: string[]
}

export interface InsertInfo {
  at: number
  offset: number
}

interface SetData<T> {
  path: string[]
  data: T
  skipHistory?: boolean
  noOverride?: boolean
  breakpoint?: string
}

interface AddColorPayload extends SetData<Record<string, unknown>> {
  id: string
}

export interface BlockInfo {
  id: string
  order: number
  top: number
  height: number
}

interface BlockState {
  type: string
  color: string
  desks: Record<string, string>
}

interface HighlightInfo {
  id: string
  height: number
}

interface ImageSwapInfo {
  path: string[]
}

export interface Desk {
  name: string
  slug: string
  url: string
}

export interface Page {
  id: string
  name: string
  slug: string
  url: string
}

export interface AddDeskPayload {
  id: string
  order: number
}

interface DataSource {
  useDataSource: (desk: Ref<Desk>, blockId: string, slotId: symbol) => Ref<Article>
}

interface SpacingConfig {
  width: number | string | Record<string, unknown>
  max?: number | string | Record<string, unknown>
  min?: number | string | Record<string, unknown>
}

export interface BlockInjected extends CommonInjected {
  desks: Desk[]
  pages: Page[]

  dataSource: DataSource

  version: number

  selectedElement: Maybe<ElementInfo>
  hoveredElement: Maybe<ElementInfo>
  selectedBlock: Maybe<BlockInfo & BlockState>
  highlightedBlock: Maybe<HighlightInfo>
  insertPoint: Maybe<InsertInfo>
  isPreview: boolean
  isPreviewHtml: boolean

  texts: Record<string, Record<string, string>>
  images: Record<string, Record<string, string>>
  blocks: string[]
  blockStates: Record<string, BlockState>

  readStyle: (path: readonly string[]) => Record<string, unknown>
  addColor: (payload: AddColorPayload) => void
  addDesk: (payload: AddDeskPayload) => void
  setSpacing: (width: SpacingConfig) => void
  setInsert: (info: Maybe<InsertInfo>) => void
  setSelectedBlock: (info: Maybe<BlockInfo>) => void
  setIsPreviewHtml: (isPreviewHtml: boolean) => void
  setElementHover: (info: Maybe<ElementInfo>) => void
  setElementSelect: (info: Maybe<ElementInfo>) => void
  setElementText: (payload: SetData<string>) => void
  setElementStyle: (payload: SetData<Record<string, unknown>>) => void
  setElementImage: (payload: SetData<string>) => void
  setImageSwapInfo: (info: Maybe<ImageSwapInfo>) => void
  uploadImage: (file: File) => Promise<string>
  setArticleCount: (blockId: string, count: number) => void
}

interface Storipress {
  clientID: string
}

export interface Injected {
  $storipress: Storipress
  $element: BlockInjected
}

type withInjected = VueConstructor<Vue & Injected>

export const INJECTED_DEFAULT: BlockInjected = {
  selectedElement: null,
  hoveredElement: null,
  scale: 1,
  version: 0,
  logo: '',
  site: { name: '', facebook: null, twitter: null },

  dataSource: {
    useDataSource: constant(
      ref({
        title: '',
        headline: '',
        desk: '',
        deskUrl: '',
        authors: [],
        blurb: '',
        url: 'https://example.com',
        time: new Date(),
      })
    ),
  },
  desks: [],
  pages: [],
  texts: {},
  images: {},
  insertPoint: null,
  blocks: [],
  blockStates: {},
  highlightedBlock: null,
  isPreview: true,
  isPreviewHtml: false,
  selectedBlock: null,
  readStyle: constant({}),
  setInsert: noop,
  addColor: noop,
  addDesk: noop,
  setSpacing: noop,
  setSelectedBlock: noop,
  setIsPreviewHtml: noop,
  setElementHover: noop,
  setElementSelect: noop,
  setElementStyle: noop,
  setElementText: noop,
  setElementImage: noop,
  setImageSwapInfo: noop,
  setArticleCount: noop,
  uploadImage: () => Promise.resolve(''),
}

export const BaseElement = (Vue as withInjected).extend({
  inject: {
    $element: {
      default: constant(INJECTED_DEFAULT),
    },
  },
})

type withElementAndID = VueConstructor<Vue & Injected & { blockId: string }>

export const BlockChild = (BaseElement as withElementAndID).extend({
  inject: ['blockId'],
})

export function useElement(): Ref<BlockInjected> {
  return ref(inject('$element', INJECTED_DEFAULT))
}
