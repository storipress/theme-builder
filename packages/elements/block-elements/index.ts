export { default as HeadlineImage } from '../common/headline-image.vue'
export { default as LinkElement } from '../common/link-element.vue'
export { default as ResponsiveImage } from '../common/responsive-image.vue'
export { default as Site } from '../common/site.vue'
export * from '../common/subscribe'
export * as limits from '../utils/limits'
export * from '../utils/width'
export { default as Article } from './article.vue'
export { default as ArticleBlock } from './article.vue'
export { default as Authors } from './authors.vue'
export type { BlockInjected, Desk } from './base-element'
export { default as Block } from './block.vue'
export { default as ColorArea } from './color-area.vue'
export { default as DeskList } from './desk-list.vue'
export { default as DeskSection } from './desk-section.vue'
export { default as FooterBlock } from './footer-block.vue'
export { default as HeroBlock } from './hero-block.vue'
export { default as Logo } from './logo.vue'
export { default as Nav } from './nav.vue'
export { default as Navigation } from './nav.vue'
export { default as PageList } from './page-list.vue'
export { default as Portal } from './portal.vue'
export { default as RawHtml } from './raw-html.vue'
export { SearchElement } from './search'
export { default as SiteInfo } from './site-info.vue'
export { default as Spacing } from './spacing.vue'
export { default as SpacingProvider } from './spacing-provider.vue'
export { default as SwappableImage } from './swappable-image.vue'
export { default as TextElement } from './text-element.vue'
export { default as TextInput } from './text-input.vue'
export { default as Icon } from 'shared/components/icon.vue'
