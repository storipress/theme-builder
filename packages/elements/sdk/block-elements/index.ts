export { default as HeadlineImage } from '../../common/headline-image.vue'
export * from '../../common/subscribe'
export * from '../common'
export { default as Article } from './article.vue'
export { default as ArticleBlock } from './article.vue'
export { default as Authors } from './authors.vue'
export { default as Block } from './block.vue'
export { default as ColorArea } from './color-area.vue'
export { default as DeskList } from './desk-list.vue'
export { default as DeskSection } from './desk-section.vue'
// in SDK, FooterBlock is just an alias for HeroBlock
export { default as FooterBlock } from './hero-block.vue'
export { default as HeroBlock } from './hero-block.vue'
export { default as Logo } from './logo.vue'
export { default as Nav } from './nav.vue'
export { default as Navigation } from './nav.vue'
export { default as PageList } from './page-list.vue'
export { default as Portal } from './portal.vue'
export { default as PortalTarget } from './portal-target.vue'
export { default as RawHtml } from './raw-html.vue'
export { default as SearchElement } from './search.vue'
export { default as SiteInfo } from './site-info.vue'
export { default as Spacing } from './spacing.vue'
export { default as SpacingProvider } from './spacing-provider.vue'
export { default as SwappableImage } from './swappable-image.vue'
export { default as TextElement } from './text-element.vue'
export { default as TextInput } from './text-input.vue'
