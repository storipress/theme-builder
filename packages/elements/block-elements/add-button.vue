<script lang="ts">
import { BaseElement } from './base-element'

export default BaseElement.extend({
  props: {
    order: {
      type: Number,
    },

    variant: {
      type: String,
      required: true,
    },
  },

  computed: {
    classes(): unknown {
      if (this.order == null) {
        return ['hidden']
      }

      const styles: Record<string, string> = (this as any).$style
      const isTop = this.variant === 'top'
      const isBottom = this.variant === 'bottom'
      const isPreview = this.$element.isPreview

      return {
        'add-button--top': isTop,
        'add-button--bottom': isBottom,
        [styles.addButton]: true,
        [styles.addButtonTop]: isTop,
        [styles.addButtonBottom]: isBottom,
        [styles.addButtonPreview]: isPreview,
        flex: !isPreview,
        hidden: !this.isBlockSelected || isPreview || this.isInsertMode,
      }
    },

    isBlockSelected(): boolean {
      return this.order === this.$element?.selectedBlock?.order
    },

    isInsertMode(): boolean {
      return !!this.$element.insertPoint
    },
  },
})
</script>

<template>
  <div :class="classes">
    <button :class="$style.addButtonBtn" @click="$emit('click')">
      <i class="icon-plus block h-6 w-6 leading-6" />
    </button>
  </div>
</template>

<style lang="scss" module>
.add-button {
  @apply absolute w-full items-center justify-center;

  :global(.block:not(.insert-mode):hover) & {
    @apply flex;
  }

  &--top {
    @apply top-0;
  }

  &--bottom {
    @apply bottom-0;
  }

  &--preview {
    @apply border-clear-blue border-b-2;
  }

  &__btn {
    @apply shadow-1 transition-colors hover:bg-gray-200;
    @apply absolute  z-10 transform rounded-full bg-white p-2;
    @apply border-light-pale-grey border shadow-sm;

    .add-button--preview & {
      @apply scale-200;
    }
  }
}
</style>
