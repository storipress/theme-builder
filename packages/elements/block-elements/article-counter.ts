import type { InjectionKey } from 'vue'
import { inject, onMounted, provide } from 'vue'
import { useElement } from './base-element'

interface ArticleCount {
  count: number
}

const ArticleCountSymbol: InjectionKey<ArticleCount> = Symbol('ArticleCount')

export function useArticleCountProvider(id: string) {
  const element = useElement()
  const articleCount = { count: 0 }

  provide(ArticleCountSymbol, articleCount)

  onMounted(() => {
    element.value.setArticleCount(id, articleCount.count)
  })
}

export function useArticleCount() {
  const articleCount = inject(ArticleCountSymbol, { count: 0 })
  articleCount.count += 1
}
