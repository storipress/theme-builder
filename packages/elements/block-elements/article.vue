<script lang="ts">
import { computed, defineComponent, inject, ref } from 'vue-demi'

import { useArticleCount } from './article-counter'
import { INJECTED_DEFAULT } from './base-element'

function filterTag(dirtyHTML: string): string {
  const parser = new DOMParser()
  const doc = parser.parseFromString(dirtyHTML, 'text/html')
  return doc.body.textContent || ''
}

export default defineComponent({
  name: 'ArticleBlock',
  setup() {
    const $element = inject('$element', INJECTED_DEFAULT)
    const deskInfo = inject('deskInfo', ref({ name: '', slug: '', url: '' }))
    const blockId = inject('blockId', '')

    const randomKey = `${blockId}_${Math.random().toString(36).slice(-8)}`
    const slotId = Symbol(randomKey)
    const article = $element.dataSource.useDataSource(deskInfo, blockId, slotId)

    useArticleCount()

    return {
      article: computed(() => {
        const { title, blurb } = article.value
        return {
          ...article.value,
          title: filterTag(title),
          blurb: filterTag(blurb),
        }
      }),
    }
  },
})
</script>

<template>
  <div>
    <!--
      @slot display article preview
      @binding {object} article article info
    -->
    <slot :article="article" />
  </div>
</template>
