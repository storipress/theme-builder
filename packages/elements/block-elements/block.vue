<script lang="ts">
import type { PropType } from 'vue-demi'
import type { BlockProp } from './use-block'
import { clickOutside } from 'shared/directives/click-outside'

import { defineComponent } from 'vue-demi'
import AddButton from './add-button.vue'
import { useArticleCountProvider } from './article-counter'
import ColorArea from './color-area.vue'
import Spacing from './spacing.vue'
import { useBlock } from './use-block'

export default defineComponent({
  components: { AddButton, ColorArea, Spacing },
  directives: { clickOutside },

  props: {
    /**
     * An required object for `Block`. Your block must accept a props called `block` and pass it to `Block`
     */
    block: {
      type: Object as PropType<BlockProp>,
      required: true,
    },

    backgroundColor: {
      type: [String, Object],
    },

    full: {
      type: Boolean,
      default: false,
    },

    isLazy: {
      type: Boolean,
      default: true,
    },
  },

  setup(props) {
    useArticleCountProvider(props.block?.id || '')
    return useBlock(props)
  },
})
</script>

<template>
  <ColorArea
    ref="root"
    v-click-outside="handleClickOutside"
    component="section"
    class="storipress-block relative transition-shadow duration-150"
    :class="blockClasses"
    :background-color="backgroundColor"
    :style="styles"
    @click="setSelected"
  >
    <AddButton :order="block.order" variant="top" @click="insertBefore" />
    <Spacing :full="full">
      <!-- @slot block content -->
      <slot />
    </Spacing>
    <AddButton :order="block.order" variant="bottom" @click="insertAfter" />
  </ColorArea>
</template>

<style lang="scss" scoped>
.highlight {
  transition: box-shadow 500ms ease-in-out 350ms;
}

.highlight-grow {
  box-shadow:
    0 1px 60px 0 rgb(0 0 0 / 10%),
    0 4px 40px 0 rgb(0 0 0 / 30%);
}
</style>
