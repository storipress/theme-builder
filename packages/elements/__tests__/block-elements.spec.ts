import { describe, expect, it } from 'vitest'
import * as siteVariant from '../block-elements'
import * as sdkVariant from '../sdk/block-elements'

function isComponentLike(name: string): boolean {
  return name[0].toUpperCase() === name[0] && name[1].toUpperCase() !== name[1]
}

describe('block elements', () => {
  it('should have the same properties as the site variant', () => {
    const keys = Object.keys(siteVariant).filter((key) => isComponentLike(key))
    for (const key of keys) {
      expect(sdkVariant).toHaveProperty(key)
    }
    expect(keys).toMatchSnapshot()
  })
})
