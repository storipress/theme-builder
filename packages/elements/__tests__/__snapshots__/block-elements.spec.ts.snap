// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`block elements > should have the same properties as the site variant 1`] = `
[
  "HeadlineImage",
  "LinkElement",
  "ResponsiveImage",
  "Site",
  "SubscribeEmail",
  "SubscribeForm",
  "Article",
  "ArticleBlock",
  "Authors",
  "Block",
  "ColorArea",
  "DeskList",
  "DeskSection",
  "FooterBlock",
  "HeroBlock",
  "Logo",
  "Nav",
  "Navigation",
  "PageList",
  "Portal",
  "RawHtml",
  "SearchElement",
  "SiteInfo",
  "Spacing",
  "SpacingProvider",
  "SwappableImage",
  "TextElement",
  "TextInput",
  "Icon",
]
`;
